_posts/2022-06-10-slick-intro.md

```diff
@@ -0,0 +1,377 @@
+In all the above samples, we used standard data types such as Int, String, Date, etc for the case classes. If we want to use a custom type, we need to provide an implicit converter to convert between the scala type and relevant column type. Let's try to use an enumeration field in case class. 
+
+```scala
+object StreamingProvider extends Enumeration {
+  type StreamingProviders = Value
+  val Netflix = Value("Netflix")
+  val Hulu = Value("Hulu")
+}
+final case class StreamingProviderMapping(
+  id: Long,
+  movieId: Long,
+  streamingProvider: StreamingProvider.StreamingProviders
+)
+```
+Now, let's create the mapping table for slick as:
+```scala
+class StreamingProviderMappingTable(tag: Tag)
+    extends Table[StreamingProviderMapping](tag, "StreamingProviderMapping") {
+
+  implicit val providerMapper =
+    MappedColumnType.base[StreamingProvider.StreamingProviders, String](
+      e => e.toString,
+      s => StreamingProvider.withName(s)
+    )
+
+  def id = column[Long]("id", O.PrimaryKey, O.AutoInc)
+  def movieId = column[Long]("movie_id")
+  def streamingProvider = column[StreamingProvider.StreamingProviders]("streaming_provider")
+  override def * =
+    (
+      id,
+      movieId,
+      streamingProvider
+    ) <> (StreamingProviderMapping.tupled, StreamingProviderMapping.unapply)
+}
+lazy val streamingProviderMappingTable = TableQuery[StreamingProviderMappingTable]
+```
+Here, we defined an implicit converter for _StreamingProvider_ enum. Since it is available in the scope, we can map the field with the enum. Slick will use the _providerMapper_ to convert between case class and database column.
+
+### 5.5. Generating DDL Scripts from Slick Tables
+Slick also provides a way to generate DDL scripts from the slick tables. This way, we can generate the table scripts and track the versions easily. This will also make sure that we can easily set up an empty database.
+To generate the DDL scripts, we need to first collect all the slick tables in a collection. 
+
+```scala
+val tables = Seq(movieTable, actorTable, movieActorMappingTable, streamingProviderMappingTable)
+```
+
+Then we can combine them into a slick DDL schema using:
+```scala
+val ddl: profile.DDL = tables.map(_.schema).reduce(_ ++ _)
+```
+Now, we can invoke the method to generate the scripts:
+```scala
+SlickTables.ddl.createIfNotExistsStatements.mkString(";\n")
+```
+
+## 6. Slick-Pg for Postgres
+PostgreSQL has additional powerful datatypes and features. We can use a thirdparty library [slick-pg](https://github.com/tminglei/slick-pg) to use those features in Slick with ease. We have already added the necessary dependencies in the _build.sbt_. 
+
+To use it, we need to write a custom Postgres Profile and use it instead of the slick provided _PostgresProfile_. We can mix-in the traits from slik-pg based on the required features of postgres. Let's add the support for JSON, HStore and Array datatypes.
+
+```scala
+trait CustomPostgresProfile
```

Comment:
I think we can break this down a bit:

Define a Profile for just _one_ of the new data types and explain the bits involved
- the Pg____Support data type
- the structure of the class, which fields/methods need to be overridden
- which implicits need to be added and why they are needed
- if/why we need a companion object and if this is best practice

Then add a new, more complex Profile which adds support for all types in one declaration (because the pattern is the same)